### PR TITLE
editorconfig/editorconfig#429 Remove section name, key and value length limits

### PR DESCRIPTION
**Pull request by goodhoko**
- Remove the upper limit for section name, key and value length.
- Permit implementations to define their own upper limits.
- Add minimum supported lengths each implementation must support.

### CODE REVIEW

**Comment by ppalaga** [comment]
@goodhoko I do not see a problem when looking at the changes. But to get real certainty I would like to see the changes necessary for ec4j to pass the new tests. Do you happen to have the changes already? Otherwise I can try.

**Comment by goodhoko** [comment]
@ppalaga Validating the tests against an actual implementation would certainly be awesome.

I do not have anything ready yet. So feel free to do that!

**Comment by ppalaga** [comment]
@goodhoko I hope I'll be able to do it in a couple of day.

File: parser/limits.in [hunk 1]
```diff
@@ -1,15 +1,13 @@
...
...aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa
```
Comment (cxw42) [inline parser/limits.in:9]: Please use the same name format as before, e.g, `k1024` instead of `key`.  It makes the code a bit more self-documenting.

Comment (goodhoko) [reply parser/limits.in]: Fixed. 🙏

**Review by ppalaga — APPROVED** [review]
ec4j is passing with these changes after I removed the limits https://github.com/ec4j/ec4j/pull/111

**Comment by cxw42** [comment]
ec-vim passes on my local machine after removing limits: editorconfig/editorconfig-vim#148

File: parser/limits.in [hunk 2]
```diff
@@ -1,15 +1,13 @@
...
-key=value
+; minimum supported section name length of 1024 characters
+[{test3,aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa}]
```
Comment (cxw42) [inline parser/limits.in:12]: There are only 1014 `a`s in this string, and the total length inside the brackets is 1022.  If I read https://github.com/editorconfig/specification/pull/21/files#diff-db4b047b30ffef7855b12b527860ac2cR134 correctly, there should be 1024 characters inside the brackets here, correct?

**Review by cxw42 — CHANGES_REQUESTED** [review]
One question

File: parser/limits.in (hunk 2 above)
Comment (ppalaga) [reply parser/limits.in:12]: Good catch. The old limit was tied to the length of the glob without brackets - at least that's how it was implemented in `ec4j`. I wonder if we should make this more explicit in the spec?

Comment (ppalaga) [reply parser/limits.in:12]: > I wonder if we should make this more explicit in the spec?

Reading again, I think it is clear enough.

Comment (goodhoko) [reply parser/limits.in]: Good catch indeed! I based the tests on the existing ones, but confused test3 and test4. 🤦 

I fixed this and also added note in the comment in the `limits.in`
```
; minimum supported section name length of 1024 characters (excluding [] brackets)
```

---

> Reading again, I think it is clear enough.

Agree. It directly follows from
> Section Name: the string between the beginning ``[`` and the ending ``]``.

Comment (goodhoko) [reply parser/limits.in]: @cxw42 can you take a look at this, please? I believe it should be all good now.

I also run the '[limitless branch](https://github.com/ec4j/ec4j/pull/111)' of ec4j against this branch and everything passed.  (thanks @ppalaga )

**Comment by ppalaga** [comment]
@cxw42 isn't this good to merge?

**Review by Mpdreamz — APPROVED** [review]
😸 LGTM

**Comment by cxw42** [comment]
@ppalaga I think we have enough successes that the spec change could now be put to a vote.

**Review by cxw42 — APPROVED** [review]

**Comment by goodhoko** [comment]
@cxw42 Rebased.

**Comment by xuhdev** [comment]
Should be be ready to be merged yet?

**Comment by cxw42** [comment]
@xuhdev pending editorconfig/specification#21, but other than that, I think it's probably fine.  The only thing is that I have not had a chance to investigate the Windows test failures on editorconfig-vim (my apologies!), but I agree we can proceed anyway.

**Comment by xuhdev** [comment]
Sorry for the delay -- sometimes I miss notification or forget things. I've started a polling: https://github.com/editorconfig/editorconfig-vote/issues/13

Feel free to ping me next time -- I'm never tired of reminders!

**Comment by goodhoko** [comment]
@xuhdev since https://github.com/editorconfig/specification/pull/21 is merged, this should be merged as well to keep the spec and tests in sync. .)

**Comment by xuhdev** [comment]
Thanks for your contributions! Sorry for having missed this PR earlier.

**Pull request merged**